truffle/src/main/java/org/jruby/truffle/nodes/core/BigDecimalNodes.java

```diff
@@ -148,41 +176,75 @@ public AddNode(RubyContext context, SourceSection sourceSection) {
             super(context, sourceSection);
         }
 
-        @Specialization(guards = "isRubyBigDecimal(b)")
-        public Object add(RubyBasicObject a, RubyBasicObject b, int precision) {
+        private Object addBigDecimal(RubyBasicObject a, BigDecimal b, int precision) {
```

Comment:
We probably want a `@TruffleBoundary` on these helper methods. Java's `BigDecimal` is unlikely to partially evaluate well, and even if it did we probably don't want the operations inlined each time they're used.